Support CLI-specific import file statements

stefansperber opened the pull request:
I added support for `source` (HiveCli) and `!run` (Beeline) commands. These commands allow to import and execute scripts in statements or scripts. This allows to compose and reuse HQL scripts.

Since these commands are not part of the core language I limited changes to the parts that emulate CLI-specific behaviour. Let me know what you think.

Some remarks:

(1) To keep `HiveServerContainer` agnostic to the used CLI I moved splitting scripts into statements out of the class. Afterwards, I removed the now unused `HiveServerContainer#executeScript` (it does not support `source/!run`). This breaks backwards compatibility. If deprecation is a better option commit `2d495443b8cc31ac99ca7c7fec77e4b193d78f0b` can be reverted.

(2) I noticed that [Beeline's SQLLine](https://cwiki.apache.org/confluence/display/Hive/HiveServer2+Clients#HiveServer2Clients-BeelineCommands) commands are not supported/emulated (for example, `!run` or `!quit`). If these should be supported in the future further refactorings could be done.

Here's an idea instead of string statements:
```
class Statement {

  private final StatementType type;
  private final String value;
  
  // ... constructor + getters for members omitted
  
  String[] getTokens() {
    return value.trim().split("\\s+");
  } 
}
```
```
enum StatementType {
  /* */
  HQL,
  /* supported by both CLIs */
  IMPORT_FILE,
  /* */
  HIVE_CLI_SPECIFIC_COMMAND,
  /* */
  BEELINE_SPECIFIC_COMMAND,
  
  // ...
}
```

This would also simplify my changes a bit more.

Example usage:
```
for (Statement statement : StatementsSplitter.splitStatements(commandShellEmulation.transformScript(script))) {
  if (StatementType.HQL == statement.getType()) {
    execute(statement);
  } else if (StatementType.IMPORT_FILE == statement.getType()) {
  	// ...
  }
}
```

stefansperber commented:
Test failures do not seem to be related to changes in the branch.

massdosage commented on src/main/java/com/klarna/hiverunner/builder/HiveShellBase.java:
```diff
@@ -100,7 +102,21 @@
+      if (tokens.length == 2) {
+        fileName = tokens[1];
+      }
+      execute(new File(fileName));
```
Do you want to do this if fileName is null? Probably not? Maybe move this into the above if statement. Is it OK to fail silently (i.e. do nothing) if fileName is null? When would this happen?

stefansperber replied:
I can see `fileName` be null when the script contains and error. For example, `source;`. In this case, a `NullPointerException` will be thrown ([File Javadoc](https://docs.oracle.com/javase/7/docs/api/java/io/File.html#File(java.lang.String))).

We could throw an `IllegalArgumentException` instead. This way it's consistent with the behaviour in `HiveShellBuilder#setScriptsUnderTest` when a script file cannot be loaded.

massdosage commented on src/test/java/com/klarna/hiverunner/sql/StatementsSplitterTest.java:
```diff
@@ -160,4 +161,16 @@ public void testReadUntilEndOfLine() {
+        Assert.assertEquals(expected, StatementsSplitter.splitStatements(expression));
+    }
+
 }
```
You should also have a test for empty statements and to check the trimming you do in the class under test.

stefansperber replied:
Good point. Will add one.

stefansperber replied:
I noticed that `expression` already adds whitespace before `statementC` to verify trimming. Do you think it's OK?

massdosage commented on src/test/java/com/klarna/hiverunner/builder/HiveShellBaseTest.java:
```diff
@@ -148,6 +151,42 @@ public void executeScriptNotStarted() throws IOException {
+      verify(container).executeStatement(hql);
+    }
+
     private HiveShell createHiveShell(String... keyValues) {
```
What about a test for whatever would result in a null file path ended up in the class under test?

stefansperber replied:
Another good point. Will add one to document the expected behaviour.

stefansperber replied:
My commit from yesterday and today moved the logic for getting the file to import into `CommandShellEmulation`. I've added tests for valid and malformed `source`/`!run` commands there. I'm waiting for a verdict on these changes; can then add extra tests for when there's trouble with the script file in `HiveShellBase`.

PelleUllberg commented:
Hi, 

Sorry for the inactivity but I've been on one of those nice long Swedish vacations again. I'm a bit choked at the moment, but will try to review this pull request in a week or two. 

Awesome feeling to get back to work and find new committers for HiveRunner!

Also, thanks a ton to @teabot for the hours you put into this project!

regards
/Pelle

Thoymas commented:
@stefansperber Could you merge up the latest changes from hiverunner master so we can see if Travis can build your pull-request?

stefansperber commented:
@Thoymas Merged your changes from `master`. Thanks for looking into the CI issue.

stefansperber commented:
I had another look at how `BeeLine` and `CliDriver` handle their import statements, and I found that my first implementation was making wrong assumptions. So I moved the logic for handling `source`/`!run` commands into `CommandShellEmulation`.

- The Hive emulation closely mirrors what `CliDriver` does (also, in the case of wrong command syntax). [Source](https://github.com/apache/hive/blob/0ae374a320d1cae523ba2b434800e97692507db8/cli/src/java/org/apache/hadoop/hive/cli/CliDriver.java#L137)

- Emulation for Beeline was less straightforward as `Commands#run` does not throw an exception when something goes wrong. Instead, a `false` value is returned which is later used to make the exit code. [Source](https://github.com/apache/hive/blob/master/beeline/src/java/org/apache/hive/beeline/Commands.java#L1535)

I chose to throw an `IllegalArgumentException` when we cannot get the filename in case the import statement's syntax is wrong. Not sure if this is the best solution; maybe both shell emulations should have similar error handling.

stefansperber commented:
@teabot 

(1) `StatementsSplitter`

Like you said. It's not great that Beeline gets special treatment in the generic `StatementsSplitter`.

I'm struggling to come up with a better alternative. Extract a `StatementSplitter` interface, and split the class up into base class and two shell-specific implementations? Or have a hook method that can be overridden to add Beeline-specific behaviour? Hope you have an idea. Please advise :)

(2) Changelog: Updated.

stefansperber commented on src/test/java/com/klarna/hiverunner/builder/HiveShellBaseTest.java:
```diff
@@ -148,6 +151,42 @@ public void executeScriptNotStarted() throws IOException {
+
+      shell.execute(file);
+
+      verify(container).executeStatement(hql);
```
`HiveShell` will read the file into a string, which is then executed the same way as a multi-statement script. Do you think this needs a extra testing? I'm reusing existing functionality.

stefansperber commented on src/main/java/com/klarna/hiverunner/sql/StatementsSplitter.java:
```diff
@@ -33,13 +33,15 @@
      */
     public static List<String> splitStatements(String expression) {
-        StringTokenizer tokenizer = new StringTokenizer(expression, SQL_SPECIAL_CHARS, true);
+        StringTokenizer tokenizer = new StringTokenizer(expression, SQL_SPECIAL_CHARS + BEELINE_SPECIAL_CHARS, true);
```
Added a comment for this on the pull request discussion page.

stefansperber commented on src/main/java/com/klarna/hiverunner/HiveServerContainer.java:
```diff
@@ -132,16 +132,6 @@ public TemporaryFolder getBaseDir() {
-     * Executes a hive script.
-     * @param hiveql hive script statements.
-     */
-    public void executeScript(String hiveql) {
```
Import statements can be used in statements and scripts. Both `CliDriver` and `BeeLine` will first read a script file in, split its lines into statements and then execute them individually. This way they can support their custom commands.

stefansperber replied:
From my understanding that is how Beeline and Hive CLI do it.

PelleUllberg commented on README.md:
```diff
@@ -227,7 +227,11 @@ Future work and Limitations
 ==============
 
-### __2.6.0-SNAPSHOT__
+### __2.6.1-SNAPSHOT__
```
Minor version should be bumped since we added functionality -> 2.7.0
Also remember to bump the version in the pom.xml file to 2.7.0

PelleUllberg commented on README.md:
```diff
@@ -227,7 +227,11 @@ Future work and Limitations
-### __2.6.0-SNAPSHOT__
+### __2.6.1-SNAPSHOT__
+
+Support shell-specific `source` (`hive`) and ``!run`` (`beeline`) commands. (These commands allow to import and execute scripts in statements or scripts.)
```
I think that this maybe deserves it's own section in the README

PelleUllberg commented on src/main/java/com/klarna/hiverunner/CommandShellEmulation.java:
```diff
@@ -35,6 +53,25 @@ public String transformScript(String script) {
+
     @Override
     public String transformStatement(String statement) {
       return filterFullLineComments(statement);
```
According to this http://sqlline.sourceforge.net/#sect_command_run any rows starting with '#' should be ignored

stefansperber replied:
The implementation seems to only support `!<command>` (see [BeeLine#dispatch](https://github.com/apache/hive/blob/master/beeline/src/java/org/apache/hive/beeline/BeeLine.java#L1128)).

PelleUllberg commented on src/main/java/com/klarna/hiverunner/builder/HiveShellBase.java:
```diff
@@ -100,7 +102,16 @@
     }
```
Why can't all command shell specific transformation happen in the transformScript call. The returned sql would have all the imports/!run commands etc. exploded and ready to be executed by the regular hive parser. This way it's easy to add more of the beeline commands etc without leaking outside of the emulator

PelleUllberg commented:
@stefansperber Any update on this PR?

stefansperber commented:
@PelleUllberg and everybody else: Thank you very much for taking your time to give me feedback. Apologies for the long silence! I should have replied earlier.

I'll start working on making the changes this week. Was a bit swamped with work recently.

stefansperber commented:
I've pushed some new changes. They are just sketches and I hope they can show where I'd like to take the implementation. Please ignore the individual commits; names and locations should also be improved.

When I went through the feedback from before I saw these issues:

1. I like the idea of including all scripts before we execute the statements. This leads to a much better separation of shell-specific behaviour. I don't want to duplicate the code for reading scripts from files/paths from `HiveShellBase` tho.

2. Charset support. The `HiveShellBase#execute` methods for files and paths also accept other `Charset`s. This seems difficult to support when we're handling single import statements.

Moving forward, I think it's best to close this pull request and submit two new ones:

- Refactor `HiveShell` and drop the methods that take other `Charset`s. Instead, set a (non-default) charset when `HiveShellBase` is constructed. This charset is used for reading all files. `HiveRunner` version should be bumped to `3.0.0`.

- Introduce `HiveQueryLanguageStatement` and move all code that deals with files/paths, include statements, ... into a factory. `HiveShellBase` does not know about splitting and transforming statements anymore but only handles collecting results and calling `HiveServerContainer`.

What do you think?

PelleUllberg commented on src/main/java/com/klarna/hiverunner/hql/HiveQueryLanguageStatementFactory.java:
```diff
@@ -0,0 +1,57 @@
+	public List<HiveQueryLanguageStatement> newInstanceForStatement(String statement) {
+		List<HiveQueryLanguageStatement> hqlStatements = new ArrayList<>();
+		String trimmedStatement = statement.trim();
+		if (commandShellEmulation.isImportFileStatement(trimmedStatement)) {
```
I still think this will become very bloated if more commands are implemented.
There should be a clean separation between implementations of the different shells and the specific commands. I rather we pass the collection of statements through the emulator and it manipulates it in whatever way needed to simulate the commands.

stefansperber commented:
Closing this for now. Unfortunately, I cannot find the time that this PR deserves at the moment. Apologies for that. Hope somebody else can base their changes on the discussion that we had and will succeed in adding this useful feature.
